crates/solver/src/in_flight_orders.rs

```diff
@@ -8,53 +34,172 @@ use std::collections::{BTreeMap, HashSet};
+    in_flight_trades: HashMap<OrderUid, PartiallyFilledOrder>,
 }
 
 impl InFlightOrders {
-    /// Takes note of the new set of solvable orders and returns the ones that aren't in flight.
+    /// Takes note of the new set of solvable orders and returns the ones that aren't in flight and
+    /// scales down partially fillable orders if there are currently orders in-flight tapping into
+    /// their executable amounts.
     pub fn update_and_filter(&mut self, auction: &mut Auction) {
         // If api has seen block X then trades starting at X + 1 are still in flight.
         self.in_flight = self
             .in_flight
             .split_off(&(auction.latest_settlement_block + 1));
 
-        // TODO - could model inflight_trades as HashMap<OrderUid, Vec<Trade>>
-        // https://github.com/cowprotocol/services/issues/123
-        // Note that this is pessimistaic as it will result in not using the
-        // remaining available amount of a partially fillable order while it is
-        // in-flight. This is done to avoid `order filled` reverts.
         let in_flight = self
             .in_flight
             .values()
             .flatten()
             .copied()
             .collect::<HashSet<_>>();
-        auction
-            .orders
-            .retain(|order| !in_flight.contains(&order.metadata.uid));
+
+        self.in_flight_trades
+            .retain(|uid, _| in_flight.contains(uid));
+
+        auction.orders.iter_mut().for_each(|order| {
+            let uid = &order.metadata.uid;
+
+            if order.creation.partially_fillable {
+                if let Some(trades) = self.in_flight_trades.get(uid) {
+                    *order = trades.order_with_remaining_amounts();
+                }
+            } else if in_flight.contains(uid) {
+                // fill-or-kill orders can only be used once and there is already a trade in flight
+                // for this one => Modify it such that it gets filtered out in the next step.
+                order.metadata.executed_buy_amount = u256_to_big_uint(&order.creation.buy_amount);
+            }
+        });
+        auction.orders.retain(|order| {
+            u256_to_big_uint(&order.creation.buy_amount) > order.metadata.executed_buy_amount
+                && u256_to_big_uint(&order.creation.sell_amount)
+                    > order.metadata.executed_sell_amount
+        });
     }
 
-    pub fn mark_settled_orders(&mut self, block: u64, orders: impl Iterator<Item = OrderUid>) {
-        self.in_flight.entry(block).or_default().extend(orders);
+    /// Tracks all in_flight orders and how much of the executable amount of partially fillable
+    /// orders is currently used in in-flight trades.
+    pub fn mark_settled_orders(&mut self, block: u64, settlement: &Settlement) {
+        let mut uids = Vec::default();
+        settlement
+            .executed_trades()
+            .inspect(|(trade, _)| {
+                uids.push(trade.order.metadata.uid);
+            })
```

Comment:
Nit: I slightly prefer doing this in an extra for loop to make it clear that we need this side effect.